### DISCUSSION TITLE
Soufflé for vim/neovim

### DISCUSSION THREAD

**Discussion by lyxell** [souffle-lang/souffle#2017] [General]
Hi! I know that there are some other Soufflé users (ping @luc-tielen) that also use vim/neovim and since the Zulip chat has been retired I think this would be a good place to discuss and share resources that are useful when developing Soufflé using vim/neovim.

To begin with there's a standard vim syntax definition found at:

* https://github.com/souffle-lang/souffle.vim

@julienhenry has also developed a tree-sitter grammar for Soufflé here:

* https://github.com/julienhenry/tree-sitter-souffle

If you are using Neovim 0.5 or later you can use this grammar to get tree-sitter based highlighting for Soufflé. You can find instructions here: https://github.com/lyxell/nvim-treesitter-souffle

If you have any other resources or tips for using Soufflé and vim together, please share!

**Comment by XiaowenHu96** [comment]
If we can have a language server for LSP that will be so cool. I started building one a few months ago but abandoned it soon after. Was quite a disaster for me to handle JSON in C++.

**Reply by lyxell** [reply]
I agree, this would be super useful! If you or someone else would want to collaborate on something I would be happy to participate.

I think @julienhenry has been working on this as well: https://github.com/julienhenry/vscode-souffle-langserver
However, I have not tested it and I am not sure how tailored that project is for vscode specifically

**Reply by luc-tielen** [reply]
LSP sounds very nice to have, but I can imagine that using C++ to build the entire thing might be a bit much.. 
Maybe it's an idea to look into embedding Lua and offloading some tasks like JSON / LSP protocol to that? And then only the logic/souffle part could be in C++.

**Comment by luc-tielen** [comment]
Thanks for the ping @lyxell! I have souffle.vim installed, but didn't know a tree-sitter grammar existed for it already, awesome!
So the way I understand it is, I need to install that grammar, then also copy your files into nvim-treesitter install location (next to highlighting queries for other langs)? Might be good to write that in the README 😄

**Reply by lyxell** [reply]
Good idea! :smile:  Anywhere in your runtimepath should be fine. I have the queries in `~/.config/nvim/queries/souffle/highlights.scm`.

**Reply by lyxell** [reply]
I fixed the directory paths to simplify installation and updated the installation instructions now. Let me know if you stumble upon any issues.

**Reply by luc-tielen** [reply]
So I have it installed, but not everything seems to be working.. like treesitter playground plugin for example. Makes me wonder if I have the plugins correctly installed?

**Reply by lyxell** [reply]
Treesitter playground should work.
Does it help if you add `autocmd BufNewFile,BufRead *.dl setfiletype souffle` to your `init`-file?
Have you disabled the classic vim-syntax definition for Soufflé? Do you get syntax highlighting at all?

**Reply by luc-tielen** [reply]
Ah yes, it was seeing my .dl files as `prolog` filetype, setting it to `souffle` works, awesome! Thanks!

**Reply by lyxell** [reply]
Great, I'll add that to the README

**Comment by b-scholz** [comment]
I am not sure whether this helps but we have started to document Soufflé's language using EBNF and rail-road grammars. 
https://github.com/souffle-lang/souffle/issues/1970

Have a look (under the Language section):
https://souffle-lang.github.io/relations
